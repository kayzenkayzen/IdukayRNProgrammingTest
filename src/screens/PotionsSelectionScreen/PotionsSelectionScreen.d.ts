export interface IPotionsSelectionScreenProps {
  navigation: any;
  potions: any;
  removePotion: any;
  addPotion: any;
  calculateOptimalAttacks?: any;
  attacks: string[][];
  damage: number;
  reset: () => void;
}
