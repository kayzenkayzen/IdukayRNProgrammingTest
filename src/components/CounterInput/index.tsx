import React, { useCallback } from 'react';
import { View, Text, TouchableOpacity, StyleSheet } from 'react-native';
import { theme } from '@src/theme';
import { ICounterInput } from './index.d';

const styles = StyleSheet.create({
  container: { flexDirection: 'row', justifyContent: 'center', height: 40 },
  button: {
    width: 40,
    backgroundColor: '#000',
    justifyContent: 'center',
    alignItems: 'center',
    borderWidth: 1,
    borderRadius: 4,
  },
  buttonLabel: {
    fontFamily: theme.fonts.primary.bold,
    fontSize: 16,
  },
  leftButton: {
    borderBottomLeftRadius: 20,
    borderTopLeftRadius: 20,
  },
  rightButton: {
    borderBottomEndRadius: 20,
    borderTopEndRadius: 20,
  },
  value: {
    width: 60,
    backgroundColor: '#fff',
    justifyContent: 'center',
    alignItems: 'center',
    borderWidth: 1,
    borderColor: '#000',
    borderRadius: 4,
    marginHorizontal: 5,
  },
  valueLabel: {
    fontFamily: theme.fonts.primary.bold,
    fontSize: 16,
    color: '#000',
  },
});

const CounterInput: React.FC<ICounterInput> = React.memo(props => {
  const {
    id,
    buttonBackgroundColor,
    buttonLabelColor,
    buttonBorderColor,
    value,
    min,
    max,
    onPressMinus,
    onPressMore,
  } = props;

  const handleOnPressMinus = useCallback(() => {
    typeof min === 'undefined' || (typeof min !== 'undefined' && value > min)
      ? onPressMinus(id)
      : null;
  }, [onPressMinus, id, min, value]);

  const handleOnPressMore = useCallback(() => {
    typeof max === 'undefined' || (typeof max !== 'undefined' && value < max)
      ? onPressMore(id)
      : null;
  }, [onPressMore, id, max, value]);

  return (
    <View style={styles.container}>
      <TouchableOpacity
        testID="minus-button"
        style={[
          styles.button,
          styles.leftButton,
          {
            backgroundColor: buttonBackgroundColor,
            borderColor: buttonBorderColor,
          },
        ]}
        activeOpacity={1}
        onPress={handleOnPressMinus}
      >
        <Text
          style={[
            styles.buttonLabel,
            {
              color:
                typeof min === 'undefined' ||
                (typeof min !== 'undefined' && value > min)
                  ? buttonLabelColor
                  : buttonLabelColor + '30',
            },
          ]}
        >
          -
        </Text>
      </TouchableOpacity>

      <View
        style={[
          styles.value,
          {
            borderColor: buttonBorderColor,
          },
        ]}
      >
        <Text testID="value-label" style={[styles.valueLabel]}>
          {value}
        </Text>
      </View>

      <TouchableOpacity
        testID="more-button"
        style={[
          styles.button,
          styles.rightButton,
          {
            backgroundColor:
              typeof max === 'undefined' ||
              (typeof max !== 'undefined' && value < max)
                ? buttonBackgroundColor
                : buttonBackgroundColor + '40',
            borderColor:
              typeof max === 'undefined' ||
              (typeof max !== 'undefined' && value < max)
                ? buttonBorderColor
                : buttonBackgroundColor + '40',
          },
        ]}
        activeOpacity={0.8}
        onPress={handleOnPressMore}
      >
        <Text style={[styles.buttonLabel, { color: buttonLabelColor }]}>+</Text>
      </TouchableOpacity>
    </View>
  );
});

CounterInput.defaultProps = {
  buttonBackgroundColor: 'transparent',
  buttonBorderColor: '#000000',
  buttonLabelColor: '#000000',
};

export default CounterInput;
