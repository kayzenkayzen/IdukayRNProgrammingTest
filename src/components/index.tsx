export { default as CounterInput } from '@src/components/CounterInput';
