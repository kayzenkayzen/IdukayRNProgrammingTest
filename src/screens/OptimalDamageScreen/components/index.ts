export { default as AttackCombinationItem } from './AttackCombinationItem/AttackCombinationItem';
