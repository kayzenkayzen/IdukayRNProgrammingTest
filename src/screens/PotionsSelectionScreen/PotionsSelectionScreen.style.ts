import { StyleSheet } from 'react-native';
import { theme } from '@src/theme';

export const styles = StyleSheet.create({
  titleContainer: {
    paddingHorizontal: 10,
    paddingVertical: 20,
    backgroundColor: '#fff',
  },
  titleText: {
    fontFamily: theme.fonts.primary.regular,
    fontSize: 16,
    lineHeight: 21,
    color: theme.colors.primary.medium,
  },
});
