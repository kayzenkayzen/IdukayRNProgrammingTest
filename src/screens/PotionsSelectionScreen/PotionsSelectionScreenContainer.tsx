import React from 'react';
import { connect } from 'react-redux';
import { OptimalDamageCalculatorActions } from '@src/redux/actions';
import { IRootState } from '@src/redux/reducers/index.d';
import { PotionsSelectionScreen } from '@src/screens';
import { IPotionsSelectionScreenProps } from './PotionsSelectionScreen.d';
import { Dispatch } from 'redux';

const StoreScreenContainer: React.FC<IPotionsSelectionScreenProps> = React.memo(
  (props: IPotionsSelectionScreenProps) => {
    return <PotionsSelectionScreen {...props} />;
  },
);

const mapStateToProps = ({ optimalDamageCalculator }: IRootState) => ({
  potions: optimalDamageCalculator.potions,
  attacks: optimalDamageCalculator.attacks,
  damage: optimalDamageCalculator.damage,
});

const mapDispatchToProps = (dispatch: Dispatch) => {
  return {
    removePotion: (potionId: string) =>
      dispatch(OptimalDamageCalculatorActions.removePotion(potionId)),
    addPotion: (potionId: string) =>
      dispatch(OptimalDamageCalculatorActions.addPotion(potionId)),
    calculateOptimalAttacks: (potions: { [x: string]: number }) =>
      dispatch(OptimalDamageCalculatorActions.calculateOptimalAttacks(potions)),
    reset: () => dispatch(OptimalDamageCalculatorActions.reset()),
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(StoreScreenContainer);
