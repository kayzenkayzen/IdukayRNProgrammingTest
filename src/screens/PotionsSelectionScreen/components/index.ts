export { default as PotionOptionItem } from './PotionOptionItem/PotionOptionItem';
