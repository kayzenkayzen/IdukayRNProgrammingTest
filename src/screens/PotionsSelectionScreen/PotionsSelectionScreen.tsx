import React, { useCallback, useEffect } from 'react';
import {
  StatusBar,
  SafeAreaView,
  ScrollView,
  View,
  Text,
  TouchableOpacity,
} from 'react-native';
import { theme } from '@src/theme';
import { constants } from '@src/constants';
import { PotionOptionItem } from './components';
import { IPotionsSelectionScreenProps } from './PotionsSelectionScreen.d';
import { styles } from './PotionsSelectionScreen.style';

const PotionsSelectionScreen: React.FC<IPotionsSelectionScreenProps> = React.memo(
  ({
    navigation,
    potions,
    removePotion,
    addPotion,
    calculateOptimalAttacks,
    attacks,
    damage,
    reset,
  }: IPotionsSelectionScreenProps) => {
    const handleOnPressCalculateOptimalAttacks = useCallback(() => {
      calculateOptimalAttacks(potions);
    }, [calculateOptimalAttacks, potions]);

    const handleRemovePotion = useCallback(
      potionId => {
        removePotion(potionId);
      },
      [removePotion],
    );

    const handleAddPotion = useCallback(
      potionId => {
        addPotion(potionId);
      },
      [addPotion],
    );

    useEffect(() => {
      if (attacks.length > 0 && damage > 0) {
        navigation.navigate('OptimalDamageScreen', {
          attacks,
          damage,
        });

        reset();
      }
    }, [attacks, damage, navigation, reset]);

    return (
      <SafeAreaView style={theme.container.outer}>
        <StatusBar
          barStyle={theme.statusBar.barStyle}
          backgroundColor={theme.statusBar.backgroundColor}
        />

        <ScrollView>
          <View style={styles.titleContainer}>
            <Text style={styles.titleText}>
              Seleccione las pociones que posee para calcular los ataques más
              efectivos
            </Text>
          </View>

          {constants.POTIONS.map((potion, index) => (
            <PotionOptionItem
              key={potion.id}
              index={index}
              id={potion.id}
              rgb={potion.rgb}
              label={potion.label}
              value={potions[potion.id]}
              onPressRemovePotion={handleRemovePotion}
              onPressAddPotion={handleAddPotion}
            />
          ))}

          {!!constants.POTIONS.reduce(
            (accumulator, currentValue) =>
              accumulator + potions[currentValue.id],
            0,
          ) && (
            <TouchableOpacity
              testID="calculate-optimal-attacks-button"
              style={theme.buttons.primary.container}
              activeOpacity={0.7}
              onPress={handleOnPressCalculateOptimalAttacks}
            >
              <Text style={theme.buttons.primary.label}>
                Calcular ataques óptimos
              </Text>
            </TouchableOpacity>
          )}
        </ScrollView>
      </SafeAreaView>
    );
  },
);

export default PotionsSelectionScreen;
