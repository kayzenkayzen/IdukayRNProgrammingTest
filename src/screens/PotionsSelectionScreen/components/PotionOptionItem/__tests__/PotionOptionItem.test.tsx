import React from 'react';
import { render } from '@testing-library/react-native';
import { PotionOptionItem } from '@src/screens/PotionsSelectionScreen/components';

describe('Test PotionOptionItem Component', () => {
  it('renders correctly', () => {
    const onPressAddPotion = jest.fn();
    const onPressRemovePotion = jest.fn();

    const wrapper = render(
      <PotionOptionItem
        index={0}
        id="red"
        rgb="red"
        label="Red Potion"
        value={0}
        onPressAddPotion={onPressRemovePotion}
        onPressRemovePotion={onPressAddPotion}
      />,
    ).toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  it('renders props correctly', () => {
    const onPressAddPotion = jest.fn();
    const onPressRemovePotion = jest.fn();

    const { getByTestId } = render(
      <PotionOptionItem
        index={0}
        id="red"
        rgb="red"
        label="Red Potion"
        value={0}
        onPressAddPotion={onPressRemovePotion}
        onPressRemovePotion={onPressAddPotion}
      />,
    );

    expect(getByTestId('potion-option-item-label').props.children).toBe(
      'Red Potion',
    );
  });
});
