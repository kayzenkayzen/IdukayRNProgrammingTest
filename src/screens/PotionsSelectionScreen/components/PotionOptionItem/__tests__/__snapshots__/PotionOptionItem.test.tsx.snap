// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test PotionOptionItem Component renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "borderBottomColor": "#dddddd",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "marginVertical": 2,
        "paddingHorizontal": 10,
        "paddingVertical": 10,
      },
      Object {
        "borderTopColor": "#ddd",
        "borderTopWidth": 1,
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/assets/images/potion.png",
        }
      }
      style={
        Array [
          Object {
            "height": 20,
            "marginRight": 10,
            "width": 20,
          },
          Object {
            "tintColor": "red",
          },
        ]
      }
    />
    <Text
      style={
        Object {
          "color": "red",
          "fontFamily": "SourceSansPro-SemiBold",
        }
      }
      testID="potion-option-item-label"
    >
      Red Potion
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "height": 40,
          "justifyContent": "center",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderBottomLeftRadius": 20,
            "borderColor": "#c0c0c0",
            "borderRadius": 4,
            "borderTopLeftRadius": 20,
            "borderWidth": 1,
            "justifyContent": "center",
            "opacity": 1,
            "width": 40,
          }
        }
        testID="minus-button"
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "SourceSansPro-Bold",
                "fontSize": 16,
              },
              Object {
                "color": "#2b2b2b30",
              },
            ]
          }
        >
          -
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#fff",
              "borderColor": "#000",
              "borderRadius": 4,
              "borderWidth": 1,
              "justifyContent": "center",
              "marginHorizontal": 5,
              "width": 60,
            },
            Object {
              "borderColor": "#c0c0c0",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000",
                "fontFamily": "SourceSansPro-Bold",
                "fontSize": 16,
              },
            ]
          }
          testID="value-label"
        >
          0
        </Text>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderBottomEndRadius": 20,
            "borderColor": "#c0c0c0",
            "borderRadius": 4,
            "borderTopEndRadius": 20,
            "borderWidth": 1,
            "justifyContent": "center",
            "opacity": 1,
            "width": 40,
          }
        }
        testID="more-button"
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "SourceSansPro-Bold",
                "fontSize": 16,
              },
              Object {
                "color": "#2b2b2b",
              },
            ]
          }
        >
          +
        </Text>
      </View>
    </View>
  </View>
</View>
`;
